icicle/backend/cpu/src/hash/cpu_poseidon.cpp

```diff
@@ -1,47 +1,307 @@
+      // Note that sparse_matrix is organized in memory 1st column first and then rest of the members of the 1st row.
+      tmp_col_res[0] = vec_in[0] * matrix_in[0];
+      for (int col_idx = 1; col_idx < T; col_idx++) { // Calc first column result.
+        tmp_col_res[0] = tmp_col_res[0] + vec_in[col_idx] * matrix_in[col_idx];
+      }
+      for (int col_idx = 1; col_idx < T; col_idx++) { // Calc rest columns results.
+        tmp_col_res[col_idx] = vec_in[0] * matrix_in[T + col_idx - 1] + vec_in[col_idx];
+      }
+      for (int col_idx = 0; col_idx < T; col_idx++) { // This copy is needed because vec_in and result storages are
+                                                      // actually the same storage when calling to the function.
+        result[col_idx] = tmp_col_res[col_idx];
+      }
+    }
+
+    void field_vec_sqr_full_matrix_mul(const S* vec_in, const S* matrix_in, S* result) const
+    {
+      unsigned int T = m_is_domain_tag ? m_arity + 1 : m_arity;
+      S tmp_col_res[T]; // Have to use temp storage because vec_in and result are the same storage.
+      for (int col_idx = 0; col_idx < T; col_idx++) { // Columns of matrix.
+        tmp_col_res[col_idx] = S::from(0);
+        for (int row_idx = 0; row_idx < T; row_idx++) { // Matrix rows but also input vec columns.
+          tmp_col_res[col_idx] = tmp_col_res[col_idx] + vec_in[row_idx] * matrix_in[row_idx * T + col_idx];
+        }
+      }
+      for (int col_idx = 0; col_idx < T; col_idx++) { // This copy is needed because vec_in and result storages are
+                                                      // actually the same storage when calling to the function.
+        result[col_idx] = tmp_col_res[col_idx];
+      }
+    }
+
+    void full_rounds(const unsigned int nof_full_rounds, S* in_out_fields, /* const */ S*& rounds_constants) const
+    {
+      unsigned int T = m_is_domain_tag ? m_arity + 1 : m_arity;
+      unsigned int alpha = poseidon_constants[T].alpha;
+      S* mds_matrix = poseidon_constants[T].mds_matrix;
+      for (int full_rounds_idx = 0; full_rounds_idx < nof_full_rounds - 1;
+           full_rounds_idx++) { // Note that the number of full rounds is nof_full_rounds-1 because of the
+                                // pre_matrix round (last round of upper full rounds) and the last round of
+                                // the bottom fulld rounds that doesn'T have sbox operation.
+        for (int state_idx = 0; state_idx < T; state_idx++) {
+          // S box
+          in_out_fields[state_idx] = S::pow(in_out_fields[state_idx], alpha);
+        }
+        for (int state_idx = 0; state_idx < T; state_idx++) {
+          // Add round constants
+          in_out_fields[state_idx] = in_out_fields[state_idx] + *rounds_constants++;
+        }
+        // Multiplication by matrix
+        field_vec_sqr_full_matrix_mul(in_out_fields, mds_matrix, in_out_fields);
+      }
+    }
+
+  private:
```

Comment:
only hash is public
all the rest should be private

Reply:
Done.